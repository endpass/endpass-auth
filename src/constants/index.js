export const METHODS = {
  SIGN: 'SIGN',
  ACCOUNT: 'ACCOUNT',
  RECOVER: 'RECOVER',
  GET_SETTINGS: 'GET_SETTINGS',

  AUTH: 'AUTH',
  LOGOUT: 'LOGOUT',
  DIALOG_RESIZE: 'DIALOG_RESIZE',
  DIALOG_OPEN: 'DIALOG_OPEN',
  DIALOG_CLOSE: 'DIALOG_CLOSE',
  INITIATE: 'INITIATE',
  READY_STATE_BRIDGE: 'READY_STATE_BRIDGE',
};

export const LS_SETTINGS = 'endpass-connect:settings';

export const IDENTITY_MODE = Object.freeze({
  DEFAULT: 'default',
  CUSTOM: 'custom',
  LOCAL: 'local',
});

export const DIRECTION = Object.freeze({
  AUTH: 'auth',
  CONNECT: 'connect',
});

export const ORIGIN_HOST = (() => {
  const parser = document.createElement('a');
  parser.href = document.referrer;
  return parser.origin;
})();
