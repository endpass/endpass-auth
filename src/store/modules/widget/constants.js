export const WIDGET_RESIZE_DURATION = 300;

export default {};
