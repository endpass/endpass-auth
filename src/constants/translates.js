import TranslateObjectFactory from '@/class/TranslateObjectFactory';

import { DOC_TYPES, DOC_STATUSES, PUBLIC_SCOPES } from '@/constants';

const { $t, createGetters } = TranslateObjectFactory;

export const DOC_TYPES_TRANSLATES = createGetters({
  [DOC_TYPES.PASSPORT]: $t('constants.documentType.passport'),
  [DOC_TYPES.DRIVER_LICENSE]: $t('constants.documentType.driverLicense'),
  [DOC_TYPES.PROOF_OF_ADDRESS]: $t('constants.documentType.proofOfAddress'),
  [DOC_TYPES.ID_CARD]: $t('constants.documentType.idCard'),
  [DOC_TYPES.SELFIE]: $t('constants.documentType.selfie'),
});

export const DOC_STATUSES_TRANSLATES = createGetters({
  [DOC_STATUSES.DRAFT]: $t('constants.documentStatus.draft'),
  [DOC_STATUSES.RECOGNITION]: $t('constants.documentStatus.recognition'),
  [DOC_STATUSES.PENDING_REVIEW]: $t('constants.documentStatus.pendingReview'),
  [DOC_STATUSES.NOT_READABLE]: $t('constants.documentStatus.notReadable'),
  [DOC_STATUSES.NOT_VERIFIED]: $t('constants.documentStatus.notVerified'),
  [DOC_STATUSES.VERIFIED]: $t('constants.documentStatus.verified'),
});

export const SCOPES_TRANSLATES = createGetters({
  // User
  [PUBLIC_SCOPES.USER]: $t('components.scopes.user.title'),
  [PUBLIC_SCOPES.USER_EMAIL_READ]: $t('components.scopes.user.email'),
  [PUBLIC_SCOPES.USER_ADDRESS_READ]: $t('components.scopes.user.address'),
  [PUBLIC_SCOPES.USER_PHONE_READ]: $t('components.scopes.user.phone'),
  // ID card
  [PUBLIC_SCOPES.DOCUMENTS_ID_CARD]: $t('components.scopes.idCard.title'),
  [PUBLIC_SCOPES.DOCUMENTS_ID_CARD_IMAGE_READ]: $t(
    'components.scopes.idCard.image',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_ID_CARD_STATUS_READ]: $t(
    'components.scopes.idCard.status',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_ID_CARD_DATA_READ]: $t(
    'components.scopes.idCard.data',
  ),
  // Passport
  [PUBLIC_SCOPES.DOCUMENTS_PASSPORT]: $t('components.scopes.passport.title'),
  [PUBLIC_SCOPES.DOCUMENTS_PASSPORT_IMAGE_READ]: $t(
    'components.scopes.passport.image',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PASSPORT_STATUS_READ]: $t(
    'components.scopes.passport.status',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PASSPORT_DATA_READ]: $t(
    'components.scopes.passport.data',
  ),
  // Driver license
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE]: $t(
    'components.scopes.driverLicense.title',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE_IMAGE_READ]: $t(
    'components.scopes.driverLicense.image',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE_STATUS_READ]: $t(
    'components.scopes.driverLicense.status',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE_DATA_READ]: $t(
    'components.scopes.driverLicense.data',
  ),
  // Proof of address
  [PUBLIC_SCOPES.DOCUMENTS_PROOF_OF_ADDRESS]: $t(
    'components.scopes.proofOfAddress.title',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PROOF_OF_ADDRESS_IMAGE_READ]: $t(
    'components.scopes.proofOfAddress.image',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PROOF_OF_ADDRESS_STATUS_READ]: $t(
    'components.scopes.proofOfAddress.status',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE_DATA_READ]: $t(
    'components.scopes.proofOfAddress.data',
  ),
});

export const SCOPES_TOOLTIPS_TRANSLATES = createGetters({
  [PUBLIC_SCOPES.USER]: $t('components.consentProvider.tooltips.user'),
  [PUBLIC_SCOPES.DOCUMENTS_ID_CARD]: $t(
    'components.consentProvider.tooltips.idCard',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PASSPORT]: $t(
    'components.consentProvider.tooltips.passport',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_DRIVER_LICENSE]: $t(
    'components.consentProvider.tooltips.driverLicense',
  ),
  [PUBLIC_SCOPES.DOCUMENTS_PROOF_OF_ADDRESS]: $t(
    'components.consentProvider.tooltips.proofOfAddress',
  ),
});
