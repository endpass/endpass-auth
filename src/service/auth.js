// @ts-check
import get from 'lodash/get';
import request from '@/class/singleton/request';
import { AUTH_STATUS_CODE } from '@/constants';

const identityBaseUrl = ENV.VUE_APP_IDENTITY_API_URL;

const TIMEOUT_DEFAULT = 1500;

/**
 * @type {object}
 */
const CODE_TO_STATUS = {
  200: AUTH_STATUS_CODE.LOGGED_IN,
  401: AUTH_STATUS_CODE.NOT_LOGGED,
  403: AUTH_STATUS_CODE.NEED_PERMISSION,
};

/**
 * @param {Promise} promise
 * @return {Promise<any>}
 */
const withSuccess = promise => {
  return promise.then(res => {
    if (res && !res.success) throw new Error(res.message);

    return res;
  });
};

/**
 * @param {Function} handler
 * @return {number}
 */
const createTimeout = handler => setTimeout(handler, TIMEOUT_DEFAULT);

/**
 * @param {string} email
 * @return {Promise<any>}
 */
const getAuthChallenge = email => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth`, {
      email,
    }),
  );
};

/**
 * @param {string} password
 * @param {string} originHost
 * @return {Promise<any>}
 */
const setAuthPermission = async (password, originHost) => {
  const res = await request.post(`${identityBaseUrl}/auth/permission`, {
    password,
    originHost,
  });
  return res;
};

/**
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.code
 * @param {string} params.password
 * @param {boolean} params.isSignUp
 * @param {boolean} params.isRemember
 * @return {Promise<any>}
 */
const authWithCode = async ({
  email,
  code,
  password,
  isSignUp,
  isRemember,
}) => {
  const url = isSignUp
    ? `${identityBaseUrl}/auth/signup`
    : `${identityBaseUrl}/auth/token`;

  return withSuccess(
    request.post(url, {
      email,
      code,
      password,
      remember: isRemember,
    }),
  );
};

/**
 * @param {string} idToken
 * @return {Promise<any>}
 */
const authWithGoogle = idToken =>
  request
    .post(`${identityBaseUrl}/auth/google`, {
      token: idToken,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

/**
 * @param {string} code
 * @return {Promise<any>}
 */
const authWithGitHub = code =>
  request
    .post(`${identityBaseUrl}/auth/github`, {
      code,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

const logout = () => request.post(`${identityBaseUrl}/logout`);

/**
 * @return {Promise<{expiresAt: number, status: string, hash: string}>}
 */
const getAuthStatus = async () => {
  try {
    const { expiresAt, hash } = await request.get(
      `${identityBaseUrl}/auth/check`,
    );
    return {
      status: AUTH_STATUS_CODE.LOGGED_IN,
      hash,
      expiresAt,
    };
  } catch (e) {
    const statusCode = get(e, ['response', 'status']);
    const status = CODE_TO_STATUS[statusCode] || AUTH_STATUS_CODE.NOT_LOGGED;

    const hash = get(e, ['response', 'data', 'hash'], '');
    const expiresAt = get(e, ['response', 'data', 'expiresAt'], 0);
    return {
      status,
      hash,
      expiresAt,
    };
  }
};

/**
 * @return {Promise<number>}
 */
const waitLogin = () =>
  new Promise((resolve, reject) => {
    /* eslint-disable-next-line */
    const handler = async function() {
      try {
        const { status } = await getAuthStatus();

        if (
          status === AUTH_STATUS_CODE.LOGGED_IN ||
          status === AUTH_STATUS_CODE.NEED_PERMISSION
        ) {
          return resolve();
        }

        createTimeout(handler);
      } catch (err) {
        return reject(err);
      }
    };

    return handler();
  });

/**
 * @param {string} email
 * @return {Promise<string>}
 */
const getRecoveryIdentifier = email =>
  request
    .get(`${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`)
    .then(res => {
      if (!res.success) throw new Error(res.message);

      return res.message;
    });

/**
 * @param {string} email
 * @return {Promise<number>}
 */
const sendEmailCode = async email => {
  const { timeout } = await request.post(`${identityBaseUrl}/auth/code`, {
    email,
  });

  return timeout;
};

/**
 * Send sms with code for disabling otp
 * @param {string} email
 * @returns {Promise<any>}
 */
const sendOtpRecoverSms = async email => {
  try {
    return withSuccess(
      request.get(
        `${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`,
      ),
    );
  } catch (error) {
    const { response = {} } = error;
    const { status } = response;
    error.code = status;

    throw error;
  }
};

/**
 * Disable otp setting
 * @param {object} param
 * @param {string} param.email
 * @param {string} param.code
 * @returns {Promise<any>}
 */
const disableOtpViaSms = async ({ email, code }) => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth/recover`, {
      email,
      code,
    }),
  );
};

/**
 * @return {string}
 */
const getSeedTemplateUrl = () => `${identityBaseUrl}/auth/recover/template`;

export default {
  getAuthStatus,
  setAuthPermission,
  getAuthChallenge,
  authWithCode,
  authWithGoogle,
  authWithGitHub,
  sendEmailCode,
  logout,
  waitLogin,
  getRecoveryIdentifier,
  sendOtpRecoverSms,
  disableOtpViaSms,
  getSeedTemplateUrl,
};
