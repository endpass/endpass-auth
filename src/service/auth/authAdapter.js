import { CHALLENGE_TYPES } from '@/constants';

function getChallengeType(challenge) {
  if (challenge === 'sms') {
    return CHALLENGE_TYPES.SMS_OTP;
  }

  if (challenge === 'otp') {
    return CHALLENGE_TYPES.APP_OTP;
  }

  return CHALLENGE_TYPES.EMAIL_OTP;
}

export default function(authData) {
  return {
    ...authData,
    challenge: {
      ...authData.challenge,
      challengeType: getChallengeType(authData.challenge),
    },
  };
}
