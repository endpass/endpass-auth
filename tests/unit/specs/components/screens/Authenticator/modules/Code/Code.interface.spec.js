import { shallowMount, createLocalVue } from '@vue/test-utils';
import { email, regularPassword as password } from '@unitFixtures/auth';
import CodeInterface from '@/components/screens/Authenticator/modules/Code/Code.interface';
import CodeInteractor from '@/components/screens/Authenticator/modules/Code/Code.interactor';
import setupI18n from '@/locales/i18nSetup';
import { CHALLENGE_TYPES } from '@/constants';

const localVue = createLocalVue();
const i18n = setupI18n(localVue);

describe('CodeInterface', () => {
  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = shallowMount(CodeInterface, {
      localVue,
      i18n,
      propsData: {
        password,
        email,
        challengeType: CHALLENGE_TYPES.EMAIL_OTP,
        isSignUp: false,
      },
    });
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('CodeInterface');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render code form', () => {
      expect(wrapper.find('code-stub').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should emit complete event', () => {
      expect(wrapper.emitted().complete).toBeUndefined();

      wrapper.find(CodeInteractor).vm.$emit('complete');

      expect(wrapper.emitted().complete.length).toBe(1);
      expect(wrapper.emitted().complete[0]).toEqual([]);
    });

    it('should emit switch event', () => {
      expect(wrapper.emitted().switch).toBeUndefined();

      wrapper.find(CodeInteractor).vm.$emit('recover');

      expect(wrapper.emitted().switch.length).toBe(1);
      expect(wrapper.emitted().switch[0]).toEqual([{ to: 'recovery-code' }]);
    });
  });
});
