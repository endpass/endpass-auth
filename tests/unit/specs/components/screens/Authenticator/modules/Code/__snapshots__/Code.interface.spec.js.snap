// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeInterface render should correctly render component 1`] = `<code-stub challengetype="EMAIL_OTP" email="user@email.com" password="11111111"></code-stub>`;

exports[`CodeInterface render should render code form 1`] = `<code-stub challengetype="EMAIL_OTP" email="user@email.com" password="11111111"></code-stub>`;
